sklearn/utils/openmp_helpers.pyx

```diff
@@ -0,0 +1,37 @@
+IF SKLEARN_OPENMP_SUPPORTED:
+    cimport openmp
+    from joblib import effective_n_jobs
+
+
+cpdef _openmp_effective_n_threads(n_threads=None):
+    """Determine the effective number of threads used for parallel OpenMP calls
+
+    - For ``n_threads = None``, returns the minimum between
+      openmp.omp_get_max_threads() and joblib.effective_n_jobs(-1).
```

Comment:
I'm not sure I understand the purpose of also checking `joblib.effective_n_jobs(-1)`?

Is this for cases where you want to run code in a joblib context manager?

Reply:
It's because effective_n_jobs is more clever than cpu_count or omp_get_max_threads is some cases. I don't recall in which situations but @tomMoral should now better.

Reply:
Ok, would be nice to have a comment explaining this then

Reply:
The `loky` implementation of `cpu_count`  better handles containers and CPU affinity. (see the docstring [here](https://github.com/tomMoral/loky/blob/master/loky/backend/context.py#L104) for detailed description).


However, it might be better to call `joblib.cpu_count` here instead of `effective_n_jobs`, as it also calls `loky.cpu_count` under the hood, .

Reply:
Reading the docstring, setting `LOKY_MAX_CPU_COUNT` will affect the output of `_openmp_effective_n_threads()`.

I don't think that makes sense?

Reply:
I don't think this is a problem and it is necessary for the `loky` test suite.

Reply:
but that's a loky variable that influences openmp behaviour?

Reply:
Its a global variable to limit the number of usable CPU in application that uses `loky`. There is a low change that it is used by anyone. Even if it is set, it would make sense to also limit the number of CPU used in `openmp` as this limit is considered as a global limit on the number of CPU.

Reply:
>Its a global variable to limit the number of usable CPU in application that uses loky.

So it has nothing to do with openmp, and it's only remotely related to scikit-learn. I don't think we should rely on this.

Reply:
I agree that it would be better it is not ideal to have this environment variable impact `openmp` usage in `scikit-learn`.

However, `sklearn.utils.cpu_count` (which is `joblib.cpu_count`) seems to be the way advocated for in the doc to get the number of CPU. So I don't think we should change this in this PR. We can open a discussion about this in another PR.

Reply:
It is only used in a single place though, in `test_multi_output_classification_partial_fit_parallelism`. None of the estimators / tools rely on it.

Also the sklearn version is deprecated.

Reply:
Indeed, it the `sklearn.utils.cpu_count` is deprecated in favor of `joblib.cpu_count`, sorry for the mistake.

Note that it is used under the hood by all calls to `Parallel`, so it is still the main way to compute the number of CPU in `scikit-learn`. I still don't think we should re-implement something in this PR, but we can discuss this in a PR in `joblib` to see if we can avoid this environment variable.

Reply:
Ok, maybe Im overthinking it. I'm interested in @ogrisel's opinion on this

Reply:
What I wanted to do is just a conservative way to find the max number of cores to use. I think that `omp_get_num_procs` could overestimate it in some situations (e.g. containers although I'm not sure about that). A more robust method is joblib's effective_n_jobs. I think you should not see this as controlling nthreads through joblib but instead use joblib to gather more informations in order to set the most appropriate number of threads.

Reply:
On the other hand I agree that the fact that it depends onthe backend is a bit brittle. Tom's proposition is better.

Reply:
I like the fact that we do not use 32 openmp threads by default when running in a docker container that can only run with a 2000 mCPU quota in the cgroups. This is typically what we have on CI systems such as travis or circle: we run in docker containers with CPU quota but `openmp.omp_get_max_threads()` would naively return 32 leading to sever oversubscription that causes a significant slow down of the computation.

@NicolasHug so you would rather duplicate the code from loky just to remove the dependency of the `LOKY_MAX_CPU_COUNT` env variable? We could do that but that feels a bit overkill. I don't expect many people to mess around with `LOKY_MAX_CPU_COUNT`.